subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java

```diff
@@ -68,9 +78,13 @@ protected Value<? extends R> calculateOwnValue(ValueConsumer consumer) {
 
         R combinedUnpackedValue = combiner.apply(leftValue.getWithoutSideEffect(), rightValue.getWithoutSideEffect());
 
-        return Value.of(combinedUnpackedValue)
-            .withSideEffect(SideEffect.fixedFrom(leftValue))
-            .withSideEffect(SideEffect.fixedFrom(rightValue));
+        Value<R> value = Value.ofNullable(combinedUnpackedValue);
+        if (keepSideEffects && !value.isMissing()) {
+            value = value
+                .withSideEffect(SideEffect.fixedFrom(leftValue))
+                .withSideEffect(SideEffect.fixedFrom(rightValue));
+        }
+        return value;
```

Comment:
I don't think we need a separate `keepSideEffects` flag. The side-effect handling in `ValueSupplier.Value` is "null-proof".

I believe the only change that is required is to use the `ofNullable` factory method.
```
        return Value.ofNullable(combinedUnpackedValue)
            .withSideEffect(SideEffect.fixedFrom(leftValue))
            .withSideEffect(SideEffect.fixedFrom(rightValue));
```

This is not a performance concern for the internal use in the `calculatePresense` as there are no side effects in 99% of cases. When there are, but the value is null, they are just ignored.

Reply:
I was saving the cost of calculating `SideEffect.fixedFrom`, which theoretically happens regardless of if it is "missing". But I suppose unless it becomes a problem, this is a premature optimization.